### PR TITLE
Cache collections with new `CacheablePathsProcessor`

### LINKED ISSUE #9058: Update Carousel Macros to be cacheable
### Problem

Our curated collections pages take way too long to load.

Blocked on #9668 

#### Relevant URL(s)
https://openlibrary.org/collections

### Reproducing the bug

1. Go to any `/collections` page

* Expected behavior: The page loads quickly.
* Actual behavior: You wait.  Then you wait some more.  Finally, the page loads.

### Context

- Browser (Chrome, Safari, Firefox, etc): All
- OS (Windows, Mac, etc): All
- Logged in (Y/N): N
- Environment (prod, dev, local): prod

#### Proposal & constraints

Add caching for `/collections` pages.

### PR DESCRIPTION
**Pull request by jimchamp**
Closes #9058

Creates new `CacheablePathsProcessor` that can be used to cache Infogami pages without overriding the default Infogami handler for such pages.

This processor attempts to fetch pages from cache when the following are true:
1. The request is a `GET`
2. The request is not specially encoded data (`json`, `yml`, etc.)
3. The request's path matches one of the patterns in the `paths_and_expiries` dictionary
4. `debug` is not in the request's query string
5. The requested page's mode is `view`

The processor has been set up to cache`/collections` and `/collections/{collection_identifier}` pages.

### Technical

Cache keys will take the form:
`{web.ctx.path}.{web.ctx.lang}{query_string}`

...where `query_string` is either an empty string or an ordered query string (e.g. `?a=1&b=1&c=1`)

Adding a new path pattern and expiry time to `paths_and_expiries` will enable caching for pages matching to path pattern.

### Testing

1. Deploy this branch to a testing environment
2. Navigate to the `/collections` page
3. On page load, refresh the page

The load time on refresh should be noticeably faster than the initial page load.

### Stakeholders

@mekarpeles

### CODE REVIEW

File: openlibrary/plugins/openlibrary/processors.py [hunk 1]
```diff
@@ -5,7 +5,10 @@
 import web
 
 from openlibrary.accounts import get_current_user
+from openlibrary.core import cache
 from openlibrary.core.processors import ReadableUrlProcessor
+from openlibrary.plugins.openlibrary.home import caching_prethread
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/processors.py:10]: **Reviewer:** Is it okay to use this `caching_prethread`?

I more or less re-implemented the homepage caching in this new processor, which uses `caching_prethread`.

File: openlibrary/plugins/openlibrary/processors.py [hunk 2]
```diff
@@ -109,6 +112,91 @@ def __call__(self, handler):
...
+                self.paths_and_expiries[match],
+                prethread=caching_prethread(),
+            )
+            _page = mc()
+
+            if not _page:
+                mc(_cache='delete')
+
+            return _page
+
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/processors.py:155]: **Reviewer:** Please check the order of the conditionals that return `handler()`.

My goal is to order these such that we exit this method as quickly as possible if the page isn't being cached by this processor.

File: openlibrary/plugins/openlibrary/processors.py [hunk 3]
```diff
@@ -109,6 +112,91 @@ def __call__(self, handler):
...
+        if 'm' in i:
+            if i['m'] != 'view':
+                # Avoid caching non-view modes
+                return handler()
+            # Prevent redundant cache entries when `m=view` is passed
+            del i['m']
+
+        # Construct cache key:
+        cache_key = f'{web.ctx.path}.{web.ctx.lang}'
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/processors.py:188]: Will be something like `/collections.en` or `/collections.it`.

File: openlibrary/plugins/openlibrary/processors.py [hunk 4]
```diff
@@ -109,6 +112,91 @@ def __call__(self, handler):
...
+        if 'debug' in i:
+            # Avoid caching error pages
+            return handler()
+
+        if 'm' in i:
+            if i['m'] != 'view':
+                # Avoid caching non-view modes
+                return handler()
+            # Prevent redundant cache entries when `m=view` is passed
+            del i['m']
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/processors.py:185]: **Reviewer:** This may be totally unnecessary...

**Review by mekarpeles — CHANGES_REQUESTED** [review]
At minimum, cache should take into consideration `lang` and `printdisabled` (in he future `sfw`

**Comment by jimchamp** [comment]
Caching localized pages was already in place (we both missed that during our quick review).  Have added additional caching for print-disabled page requests, and have included commented-out code for `sfw`.

**Comment by jimchamp** [comment]
Blocking temporarily for PM script testing purposes.

**Pull request merged**